### DISCUSSION TITLE
Add mail to the context of custom endpoint/custom hooks

### DISCUSSION THREAD

**Discussion by workatease** [directus/directus#4664] [Feature Requests]
Add the mail util in the context for endpoint and hooks 
it will be easy to send mail
https://github.com/directus/directus/blob/main/api/src/mail/index.ts

**Comment by workatease** [comment]
and the logger too

**Comment by rijkvanzanten** [comment]
I think we should spend some time revamping the mail util a bit so it becomes a more powerful mail service (with custom templates etc), otherwise it'll be a breaking change down the line 🤔

**Reply by benhaynes** [reply]
I agree with this.

**Reply by benhaynes** [reply]
Reference Discussions:

https://github.com/directus/directus/discussions/3104
https://github.com/directus/directus/discussions/2774
https://github.com/directus/directus/discussions/2773
https://github.com/directus/directus/discussions/3114
https://github.com/directus/directus/discussions/3115
https://github.com/directus/directus/discussions/3465
https://github.com/directus/directus/discussions/3741

**Reply by workatease** [reply]
> Reference Discussions:
> 
> #3104
> #2774
> #2773
> #3114
> #3115
> #3465
> #3741

I should search first before I start an discussion.

**Reply by benhaynes** [reply]
Nah, I had to search for these... they're relevant, but all are a little different. I just wanted us to have the "full picture" in case someone wanted to dive into a PR. 🙂

**Comment by workatease** [comment]
> I think we should spend some time revamping the mail util a bit so it becomes a more powerful mail service (with custom templates etc), otherwise it'll be a breaking change down the line 🤔 

Yes .. I agree ... bring your own template by overriding the build in one is a good thought... I  am happy to help

I was also thinking of if we need to bring In the directus utils as part of the context

**Comment by workatease** [comment]
I was thinking like creating a template directory under extensions, 
the mail service will look for to overridden template template with the same name.
if not use the build in template

still not clear on the other requirement on how the permission will take effect based on role.

**Comment by workatease** [comment]
can we now add the utils to context now?

**Reply by rijkvanzanten** [reply]
> can we now add the utils to context now?

I'm afraid not. If we do that now, it would mean that the update to a more "full fledged" mail setup would be a breaking change 😨

**Reply by workatease** [reply]
yes I understand your concern. I was talking about the directus Utils and logger
https://github.com/directus/directus/tree/main/api/src/utils

**Reply by rijkvanzanten** [reply]
Ohh my bad!

The problem with exposing those internals to extensions is that we can't update the utils anymore afterwards 🤔 EG right now, we can make changes to those utils however often we want, because they're "internal" and thus not a breaking change. That will be complicated quite a bit once we start exposing all of them by default. 

I'd rather compile a list of the couple utils that are mission-critical to have in extensions (like `get-schema`) and only expose those 🤔

**Reply by workatease** [reply]
yes I hope the list will have async-handler as it will help in custom Endpoint hooks

**Comment by jaads** [comment]
Heya! Thank you for taking the time to submit this request! 

It has been over **90 days**, and this discussion has not received at least **15 votes** from the community. This means that we don't feel like there's enough community interest to warrant further R&D into this topic at this time. :ice_cube:

This request will now be closed to keep our discussions tidy. Please reach out if you have any questions!

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._
